mlir/lib/Target/LLVMIR/Dialect/ROCDL/ROCDLToLLVMIRTranslation.cpp

```diff
@@ -100,6 +100,12 @@ class ROCDLDialectLLVMIRTranslationInterface
         llvmFunc->addFnAttr("amdgpu-flat-work-group-size", "1,256");
       }
 
+      // MLIR's GPU kernel APIs all assume and produce uniformly-sized
+      // workgroups, so the lowering of the `rocdl.kernel` marker encodes this
+      // assumption. This assumption may be overridden by setting
+      // `rocdl.uniform_work_group_size` on a given function.
+      if (!llvmFunc->hasFnAttribute("uniform-work-group-size"))
```

Comment:
Why do you need to check if the attribute is already set on the function? Aren't you constructing this fresh IR?

Reply:
I don't have guarantees about whether we'll process `rocdl.kernel` or `rocdl.uniform_work_group_size` first, so if I've already encountered a manual definition of uniform work group size, I don't want to overwrite it with the default

Reply:
This seems like a peculiar process but I have no idea how MLIR does things. I'd expect it to look like clang where the full set of attributes is considered when initially emitting the IR attributes

Reply:
Yeah, this is a hook `amendOperation`. It gets called when you have a `rocdl` attribute on an operation being translated, once per attribute.
So given
```mlir
llvm.func @foo(...)  attributes {llvm.something = N : i64, rocdl.kernel, rocdl.uniform_work_group_size = true}
```
The LLVM dialect is responsible for translating `llvm.func` into LLVM IR, and then you'll end up with three `amendOperation()` calls in some order, one to the LLVM dialect's, two to ROCDL's.